### ISSUE TITLE
Draw.io using SVG to support links within diagram

### ISSUE THREAD

**Issue by mark-james** [BookStackApp/BookStack#1170]
**Describe the feature you'd like**
If the draw.io integration saved the diagram as SVG rather than png then links created within the diagram could also be followed directly from the bookstack page

**Describe the benefits this feature would bring to BookStack users**
Diagrams elements can be used as shortcuts to access other pages within bookstack.

Often more visual representations are easier to understand. So a diagram could replace a table of contents. 

This provides another form of interlinking between pages within bookstack.

**Additional context**
With E11 support for bookstack disappearing anyway, is there a reason not to use SVG instead?

**Comment by ssddanbrown** [comment]
Thanks for the suggestion @mark-james, I had not thought about links within drawings.

Would somewhat rely on #1103 in addition to many other changes.

> is there a reason not to use SVG instead?

Yeah, SVG's are harder to deal with, There's a lot more to them (From an editor/display point of view) whereas with a PNG everything is effectively encapsulated within a single unit.

**Comment by mark-james** [comment]
Ok. Thanks @ssddanbrown . 

I'm happy to contribute a pull request if you could point me in the right direction as far as changes that need to be made? (Or at least things I need to consider)

This is a fairly important change to me as I'm going to start moving a knowledge base onto bookstack soon. The existing solution while inferior in many ways unfortunately does allow for links within diagrams. So it's going to cause me some conflict if I can't provide users this feature.

**Comment by ssddanbrown** [comment]
### Considerations

* The back-end image upload endpoints would need to be made compatible with SVG.
* The front-end image manager would need to be updated to work with SVG.
* The draw.io front-end code would need to be updated to handle draw.io SVGs while ensuring full feature compatibility when re-importing (So no layer loss for example, Would need to check if SVG's retain all data themselves or if extra metadata needs to be saved).
* All of the above would need to be made backwards compatible with the existing png-based system.
* Both the markdown editor and WYSWIYG editor drawing implementations would need to be updated to handle SVG's, Might be a challenge to cleanly a large drawing as SVG content in markdown due to the amount of content that will be embedded. Might tie into next point:
* Need to consider how to display SVG drawings without issue, Without limited possibility of external page styles affecting them. This will be tricky as I think draw.io drawings can contain standard elements such as `<h1>` or `<p>` tags. Maybe you could iframe them into the page? Would mean cleaner markdown code but sizing/layout gets trickier. Perhaps shadow DOM could work here? Needs investigation.

**Comment by mark-james** [comment]
Thanks @ssddanbrown.
Lots to consider!
I'll begin looking at this early next week.

**Comment by bharatrajagopalan** [comment]
Hi

Since this topic is regarding svg, thought would add in my use case


As of now I  can use the markdown editor in bookstack to copy paste the xml content from an svg and paste it so that it renders on bookstack including links - i can then use the svg viewBox attribute to control how it fits/scales in the page. See screenshots




I currently use Plantuml to generate UML diagrams with SVG and these are interactive, e.g. i can click on a ref in one diagram to take me to another page - i use a common constants file to maintain the links to wiki pages and also the skin to ensure that the uml diagram looks consistent.

However the process of copy pasting svg into a page is cumbersome and if i change one link or colour in my constants, while i have a script to regenerate all my uml svgs, i do need to manually update them one by one on each equivalent wiki pae.

As a starting point, It would be great if there was a way for me to refer to svgs from an attachment rather than having to paste them inline in html - in wikijs or confluence, i can use jquery to load them into an svg tag from using the link to the attachment. Which means that when i do mass updates i simply upload all the svgs and it overwrites the older content which then gets loaded on the page - links and all.

One obvious answer is iframe but would prefer if object tag was supported 

I also noticed that when i upload svgs, and try reuploading them into bookstack it creates a new copy - it would be great if the files could simply be overriden.


Of course even better if bookstack supported plantuml directly such that i could type plantuml syntax inline which gets rendered (including the ability to include external plantuml files as skins/constants) but that is a bit of a pain as it needs graphviz/plantuml to be setup on the server with the wiki which may not always be possible.

In wikijs and confluence i can use the {html} markdown mode to use jquery to dynamically load html from an attachment into an svg tag like this

<div style="text-align:center" >
	<div id="svg" />
</div>
<script>
    $(document).ready(function(){  
      $('#svg').load("/download/attachments/56165089/005_Data_Setup_Loads.svg?api=v2");
    });
</script>

That's obviously cumbersome and a security risk and not user friendly so it would be nice if bookstack had a way to insert svgs like the way it can insert images but maintain interactiveness

<img width="1440" alt="screenshot 2018-12-15 at 12 43 49" src="https://user-images.githubusercontent.com/15735432/50043175-766d0b00-0067-11e9-8a52-aa042f7402b9.png">
<img width="1440" alt="screenshot 2018-12-15 at 12 44 26" src="https://user-images.githubusercontent.com/15735432/50043176-7705a180-0067-11e9-9baf-483dbc42af8a.png">

**Comment by simchanu29** [comment]
Draw.io can generate embeded html.
Would integrating this functionnality as html code be easier ?

**Comment by ssddanbrown** [comment]
@simchanu29 Unfortunately not really. Some considerations above would be lesser, but some new considerations would be introduced.

**Comment by bakkertjebrood** [comment]
Is there any update on this issue? Would be really nice to have links within the created diagrams.

**Comment by ssddanbrown** [comment]
@bakkertjebrood No, No update.

**Comment by passchn** [comment]
Are there any updates on this or do you plan to support SVG/HTML embeds in the future?

**Comment by ssddanbrown** [comment]
No, no further update on this. Would likely be activity here if anything was in progress.

**Comment by ssddanbrown** [comment]
Have attempted to solve this within #3452.
Unfortunately, the complexity and lack of portability of drawings as SVGs have proved to cause issue. 
Draw.io diagrams rely on web-rendering, not just SVG compliance, and are too complex for our default PDF renderer and reflects poorly on content portability in general when in SVG format.

Not sure where to take this. Could export in both formats but then requires duplicate storage and format tracking/juggling, Not a mess I want to walk into. Maintaining a choice between "SVG-based drawings or working PDF exports" would also suck.

**Comment by Matthew2000** [comment]
Is it possible to use Draw.io in a view only mode? [https://www.diagrams.net/blog/online-diagram-viewer](https://www.diagrams.net/blog/online-diagram-viewer) I don't know how exactly this works but it seems to indicate that there is a view only mode.

This might be a solution to getting links made in the diagrams to be usable and allowing viewers a better way to look at them.

**Comment by fmos** [comment]
> Have attempted to solve this within #3452. Unfortunately, the complexity and lack of portability of drawings as SVGs have proved to cause issue. Draw.io diagrams rely on web-rendering, not just SVG compliance, and are too complex for our default PDF renderer and reflects poorly on content portability in general when in SVG format.
> 
> Not sure where to take this. Could export in both formats but then requires duplicate storage and format tracking/juggling, Not a mess I want to walk into. Maintaining a choice between "SVG-based drawings or working PDF exports" would also suck.

Thank you for your effort @ssddanbrown and for providing bookstack in the first place!

In #3452 I noticed that you checked off "Support SVG for the standard image gallery." 

Would you want to consider to merge this feature, even without SVG support for embedded drawings? I guess, this would allow to create diagrams with links outside bookstack and embed them.

**Comment by ssddanbrown** [comment]
> Would you want to consider to merge this feature, even without SVG support for embedded drawings?

@fmos No, the same issues for drawings would apply for images. The referenced checked off todo item was not in relation to full supported and accepted SVG support for image uploads, it was just in reference to adding support to the image manager functions.

**Comment by vincentbernat** [comment]
Alternatively, we could double the size of the PNG. This can be done by adding `scale: 2` to the export action. We can apply 50% for width/height styles to get back the image in the right format. However, I don't know how to handle the transition. The scale should be stored somewhere.

**Comment by ssddanbrown** [comment]
@vincentbernat That wouldn't address the fundamental request here, of supporting links within drawings.
I'd consider resolution/pixel-density to be a separate desire/request/issue/conversation.

**Comment by at-ng** [comment]
Something to consider when implementing this, with SVG another nice feature that could be added at the same time is the possibility to index diagrams. That way you could find things that are in diagrams when searching (just being able use the standard browser page search would be good as well). Sometimes you need to have very large diagrams then search would be great (zoom as well but that might be a different issue).

**Comment by sarang-apps** [comment]
Could a possible solution to handle this be to use a javascript canvas with the background image as a png, and links as regions with onclick events?

**Comment by syh7** [comment]
Just bumping this to show there is still interest in allowing links in diagrams

**Comment by Bert-Proesmans** [comment]
I've been following this thread for a while and wanted to make the same suggestion as simchanu29 again, and i'm actually not sure if his suggestion was interpreted correctly as using [draw.io lightbox mode](https://github.com/jgraph/drawio/wiki/Embed-Diagrams) instead of exporting as html. Don't export the diagram as HTML.

A purely additive feature would be to keep the PNG render as main and printable content, but while in read mode add an onclick handler to the PNG that opens a css lightbox² displaying drawio in lightbox mode (\<diagram-url\>/?lightbox=1#\<diagram-data\>).

This does not allow for indexing the diagram contents since the diagram is still considered an opaque blob, just like today. BUT it does allow for diagram text to be selectable, diagram links to be clickable, diagrams to be interactive (if using draw.io plugins). And the best part is that we skip the headache that is purifying the SVG data itself because of cross-site scripting risk.

To directly answer on the first post, this approach allows
✔️ Diagrams elements can be used as shortcuts to access other pages within bookstack.

❓   Often more visual representations are easier to understand. So a diagram could replace a table of contents.
=> Opening draw.io in lightbox mode will happen through an iframe which is a css block element and difficult to precisely position into text content. But not impossible to do, with/without custom styling.
❗ Embedded SVG's would be better since text can flow around and through the diagram.
❗ I'm interested to see examples of more closely related diagrams to text for functional purposes. In my opinion this leads to a bad content state, but it's not my intention to judge.

✔️ This provides another form of interlinking between pages within bookstack.

The implementation would be similar to clicking on the PNG from article edit mode, it only uses a different diagrams 'launch mode'. ²The css lightbox is just fancy, clicking the image might as well open drawio spanning the entire viewport. If the css lightbox could be stylable after a few iterations on the code, each user can decide for themselves if they want to open the diagram in fullscreen or some other way. Fullscreen will probably be enough for most people.

**Comment by carlossierra311** [comment]
We were just creating some navigation maps for our end users and found out that all the links included on the diagram didn´t work. We would really appreciate this capability as well if it could find its way into BookStack. 

Ps. No matter the outcome of this thread/request, I'd like to take the opportunity to thank you, @ssddanbrown for all the splendid work you do to bring us such a useful tool. Thanks!!!

**Comment by hdeppert** [comment]
I just checked the two SVG options and the first one is as simple and easy as the PNG option:
1. Save the draw.io diagram as a SVG file -> http/s links are working fine in browsers, but not links to subpages  of the diagram
2. Use draw.io SVG Embedding -> all links are working, but only in the fullscreen view calling the draw.io-viewer
    - I guess this could be somehow customized to use a self-hosted viewer instance 

IMHO: For me, the most interactive / convenient way would be to use the draw.io HTML Embedding directly, fully inline with full interaction/link capabilities.

**Comment by PhilippRieth** [comment]
To get this functionality, we currently export diagrams as iframes and embed them into BookStack. The only downside is that every time you open the diagram and perform a change, you have to again, export it as an iframe again and replace the existing diagram on the BookStack page. I think this is what @Matthew2000 suggested. 

Export as iframe (with base64 encoded diagram in iframe URL)
![image](https://github.com/user-attachments/assets/1ba1e1b6-5905-45d1-85c9-02d2bd5b310b)

Embedded as media into BookStack page
![image](https://github.com/user-attachments/assets/e7577b99-8529-431c-9006-991b57f22aab)

have clickable links in BookStack (diagrams.net viewer)
![image](https://github.com/user-attachments/assets/e1b13726-ed4e-4cf5-8de9-da70ed2c1e00)


It's a workaround, but it works for us as we don't change most flow charts on a daily basis.

**Comment by ssddanbrown** [comment]
Over the last few days I've been playing exploring with hybrid approaches on this, specifically extracting drawing data from the BookStack-stored png drawings, to then display them in an interactive viewer of some kind.

I've formed my result of this into a hack here: https://www.bookstackapp.com/hacks/interactive-drawings/
This solution extracts the drawing data, and then replaces the drawings in-page with iframes which load the drawing data via the hosted “https://viewer.diagrams.net/” service. This does have some downsides/considerations as noted on that hack page.

I did attempt to load via non-iframe means for a more native embed (faster loading, better interaction for things like lightbox viewing, and potentially without having to rely on the external site for loading), but these avenues generally resulted in a more problematic result than that of the iframe embed I landed on above (mainly due to CSS pollution issues).

**Comment by at-ng** [comment]
A shame it could not be more tightly integrated. Not sure what avenues you attempted but since you mention CSS pollution, did you try using shadow DOM?

Either way, thanks for the "hack". I have not had the chance to test it yet, I assume this will not let Bookstack index the content of the diagram to make it searchable but probably works with CTRL+F in your browser for that specific page.

**Comment by ssddanbrown** [comment]
@at-ng Yeah, I did try out shadow-dom (in addition to CSS override reset options, and same-origin simple iframe scoping). The viewer isn't really designed to be rendered with the shadow-dom, so while it could render fine, interaction would be funky/broken with errors. I could avoid the main errors with some monkey-patching of browser global functions, but that did not help interaction. 

It might be possible to render within iframes using a locally hosted viewer, with specific added handling to detect things like lightbox usage (and other interaction) to then handle that (iframe positioning/sizing) to adapt, but that increases the scope of the customization quite a bit while I had already spent a while testing options, and the iframe solution landed upon above will probably help in a lot of people's cases.

> I assume this will not let Bookstack index the content of the diagram to make it searchable but probably works with CTRL+F in your browser for that specific page.

That's correct.

**Comment by ssddanbrown** [comment]
BookStack is now [managed on Codeberg](https://codeberg.org/bookstack/bookstack).

As part of this change, we're closing all existing GitHub issues and pull requests.
You'll instead find this issue on Codeberg here: 
https://codeberg.org/bookstack/bookstack/issues/1170
  
Any further discussion for this will now be done on Codeberg.
Therefore this GitHub thread will be locked.
